// @tidyjs/tidy-moment/dist/lib/summarizeMomentGranularity.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var moment = require('moment');
var tidy = require('@tidyjs/tidy');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var moment__default = /*#__PURE__*/_interopDefaultLegacy(moment);

function summarizeMomentGranularity(granularity, summarySpec, options) {
  const _summarizeMomentGranularity = (items) => {
    options = options != null ? options : {};
    const {
      timestampKey = "timestamp",
      dateKey = "date",
      rest = tidy.first,
      ...other
    } = options;
    const dateFormats = {
      s: "YYYY-MM-DDTHH:mm:ss",
      seconds: "YYYY-MM-DDTHH:mm:ss",
      second: "YYYY-MM-DDTHH:mm:ss",
      min: "YYYY-MM-DDTHH:mm",
      minutes: "YYYY-MMDDTHH:mm",
      minute: "YYYY-MM-DDTHH:mm",
      d: "YYYY-MM-DD",
      days: "YYYY-MM-DD",
      day: "YYYY-MM-DD",
      w: "GGGG-WW",
      week: "GGGG-WW",
      weeks: "GGGG-WW",
      m: "YYYY-MM",
      month: "YYYY-MM",
      months: "YYYY-MM",
      q: "YYYY-[Q]Q",
      quarter: "YYYY-[Q]Q",
      quarters: "YYYY-[Q]Q",
      y: "YYYY",
      year: "YYYY",
      years: "YYYY"
    };
    const dateFormat = dateFormats[granularity];
    const results = tidy.tidy(items, tidy.mutate({
      [dateKey]: (d) => moment__default['default'].utc(d[dateKey].format(dateFormat), dateFormat),
      [timestampKey]: (d) => d[dateKey].toISOString()
    }), tidy.groupBy([timestampKey], [
      tidy.summarize(summarySpec, {rest, ...other})
    ]));
    return results;
  };
  return _summarizeMomentGranularity;
}

exports.summarizeMomentGranularity = summarizeMomentGranularity;
